clippy_lints/src/casts/unnecessary_cast.rs

```diff
@@ -142,6 +144,33 @@ pub(super) fn check<'tcx>(
     }
 
     if cast_from.kind() == cast_to.kind() && !expr.span.in_external_macro(cx.sess().source_map()) {
+        enum MaybeParenOrBlock {
+            Paren,
+            Block,
+            Nothing,
+        }
+
+        fn is_borrow_expr(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+            matches!(expr.kind, ExprKind::AddrOf(..))
+                || cx
+                    .typeck_results()
+                    .expr_adjustments(expr)
+                    .iter()
+                    .any(|adj| matches!(adj.kind, Adjust::Borrow(_)))
```

Comment:
Another small nit: Only the first adjustment needs to be checked. Any later borrow would have to borrow a temporary.